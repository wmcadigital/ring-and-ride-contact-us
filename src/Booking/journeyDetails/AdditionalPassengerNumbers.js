import { useFormState, useForm } from "react-final-form";

import FormSection from "../../common/FormSection";
import Question from "../../common/Question";
import ProgressIndicator from "../../common/ProgressIndicator";
import NumberInput from "../../common/NumberInput";
import getSectionPositionInfo from "../getSectionPosition";
import getAboutJourneySectionName from "./getAboutJourneySectionName";
import {
  required,
  numbersOnly,
  numberGreaterThanZero,
  composeValidators,
} from "../../common/validation";

const AdditionalPassengerNumbers = () => {
  const stateApi = useFormState();
  const formApi = useForm();

  const formValues = stateApi.values;
  const bookingParty = formValues["bookingParty"];
  const passengerNumbers = formValues["additionalPassengerNumbers"];

  let question = "";
  if (bookingParty === "mySelf") {
    question = "How many people are joining you on the journey?";
  } else {
    question = `How many people are joining ${formValues["firstName"]} on the journey?`;
  }

  const pronoun = formValues["bookingParty"] === "mySelf" ? "you" : "them";

  const error = stateApi.submitFailed
    ? stateApi.errors?.["additionalPassengerNumbers"]
    : null;

  const decreaseNumberCallback = () =>
    formApi.mutators.setFormAttribute(
      "additionalPassengerNumbers",
      passengerNumbers - 1
    );
  const increaseNumberCallback = () =>
    formApi.mutators.setFormAttribute(
      "additionalPassengerNumbers",
      passengerNumbers + 1
    );

  return (
    <FormSection>
      <ProgressIndicator
        sectionPosition={getSectionPositionInfo("2", stateApi)}
        sectionName={getAboutJourneySectionName(stateApi)}
      />
      <Question text={question} />
      <NumberInput
        fieldName="additionalPassengerNumbers"
        label={`Number of people joining ${pronoun} on the journey`}
        defaultValue={1}
        validation={composeValidators(
          required,
          numbersOnly,
          numberGreaterThanZero
        )}
        decreaseNumberCallback={decreaseNumberCallback}
        increaseNumberCallback={increaseNumberCallback}
        error={error}
      />
    </FormSection>
  );
};

export default AdditionalPassengerNumbers;
