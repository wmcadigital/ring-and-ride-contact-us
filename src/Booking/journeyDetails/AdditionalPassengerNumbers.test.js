import { useFormState } from "react-final-form";
import { createRenderer } from "react-test-renderer/shallow";

import AdditionalPassengerNumbers from "./AdditionalPassengerNumbers";

jest.mock("react-final-form");

describe("AdditionalPassengerNumbers - snapshots", () => {
  it("individual is filling out form for themselves", () => {
    useFormState.mockImplementationOnce(() => ({
      values: {
        bookingParty: "mySelf",
      },
    }));
    const renderer = createRenderer();

    renderer.render(<AdditionalPassengerNumbers />);

    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });

  it("individual is filling out form for someone else", () => {
    useFormState.mockImplementationOnce(() => ({
      values: {
        bookingParty: "behalfSomeone",
        firstName: "John",
      },
    }));
    const renderer = createRenderer();

    renderer.render(<AdditionalPassengerNumbers />);

    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });

  it("renders with an error", () => {
    useFormState.mockImplementationOnce(() => ({
      values: {
        bookingParty: "mySelf",
        firstName: "John",
        additionalPassengerNumbers: -1,
      },
      submitFailed: true,
      errors: { additionalPassengerNumbers: "Invalid" },
    }));
    const renderer = createRenderer();

    renderer.render(<AdditionalPassengerNumbers />);

    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });
});
