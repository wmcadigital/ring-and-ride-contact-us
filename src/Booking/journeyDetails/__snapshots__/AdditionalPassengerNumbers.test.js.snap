// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdditionalPassengerNumbers - snapshots individual is filling out form for someone else 1`] = `
<FormSection>
  <ProgressIndicator
    sectionName="Plan their journey"
    sectionPosition="Section 2 of 4"
  />
  <Question
    text="How many people are joining John on the journey?"
  />
  <NumberInput
    decreaseNumberCallback={[Function]}
    defaultValue={1}
    error={null}
    fieldName="additionalPassengerNumbers"
    increaseNumberCallback={[Function]}
    label="Number of people joining them on the journey"
    validation={[Function]}
  />
</FormSection>
`;

exports[`AdditionalPassengerNumbers - snapshots individual is filling out form for themselves 1`] = `
<FormSection>
  <ProgressIndicator
    sectionName="Plan your journey"
    sectionPosition="Section 2 of 4"
  />
  <Question
    text="How many people are joining you on the journey?"
  />
  <NumberInput
    decreaseNumberCallback={[Function]}
    defaultValue={1}
    error={null}
    fieldName="additionalPassengerNumbers"
    increaseNumberCallback={[Function]}
    label="Number of people joining you on the journey"
    validation={[Function]}
  />
</FormSection>
`;

exports[`AdditionalPassengerNumbers - snapshots renders with an error 1`] = `
<FormSection>
  <ProgressIndicator
    sectionName="Plan your journey"
    sectionPosition="Section 2 of 4"
  />
  <Question
    text="How many people are joining you on the journey?"
  />
  <NumberInput
    decreaseNumberCallback={[Function]}
    defaultValue={1}
    error="Invalid"
    fieldName="additionalPassengerNumbers"
    increaseNumberCallback={[Function]}
    label="Number of people joining you on the journey"
    validation={[Function]}
  />
</FormSection>
`;
